File: day_05/p1.py
from typing import Dict, Generator, List, Tuple


def parse(fname: str) -> Tuple[Dict[str, List[str]], List[List[str]]]:
    lines = []
    rules: Dict[str, List[str]] = {"-1" : []}
    updates = []

    with open(fname, 'r') as f:
        lines = f.read().splitlines()
    
    for line in lines:
        if '|' in line:
            # rule
            first, second = line.split('|')
            if first in rules.keys():
                rules[first].extend([second])
            else:
                rules[first] = [second]
        elif ',' in line:
            #update
            updates.append(line.split(','))
    
    for v in rules.values():
        v.sort()
    
    return rules, updates

def gen(lst: str, idx: int) -> Generator[str, None, None]:
    yield lst[len(lst) - idx:]
    
def compute(rules: Dict[str, List[str]], updates: List[str]) -> int:
    bad_updates = set()
    good_updates = set()

    for update in updates:
        for i, page in enumerate(update[::-1][1:]):
            i += 1
            # for each page, check that none of the following ones should come first
            for elem in gen(update, i):
                for e in elem:
                    if e in rules.keys() and page in rules[e]:
                        bad_updates.add(tuple(update))
                    else:
                        good_updates.add(tuple(update))
    
    good_updates = {tuple(update) for update in updates} - bad_updates

    return sum (int(good[len(good) // 2]) for good in good_updates)

def main():
    rules, updates = parse("input.txt")
    print(compute(rules, updates))

if __name__ == "__main__":
    main()

File: day_05/p1.cpp
#include "pypp.hpp" // https://github.com/A-Bergonzoli/pypp/tree/main

#include <iostream>
#include <unordered_map>

using strings = std::vector<std::string>;
using ints = std::vector<int>;
using mmap = std::unordered_map<std::string, ints>;

std::pair<mmap, std::vector<ints>> Parse(const std::string& fname)
{
    mmap rules {};
    std::vector<ints> updates {};
    const auto lines = pypp::splitFileLines(fname);

    for (const auto& line : lines) {
        if (line.find('|') != std::string::npos) { // rule
            const auto first = pypp::splitThenGetAt(line, '|', 0);
            const auto second = pypp::splitThenGetAt(line, '|', 1);
            rules[first].push_back(std::stoi(second));
        } else if (line.find(',') != std::string::npos) { // update
            auto updates_strings = pypp::split(line, ',');
            ints update {};
            std::ranges::transform(updates_strings,
                std::back_inserter(update),
                [](const std::string& s) { return std::stoi(s); });
            updates.push_back(update);
        }
    }

    return std::make_pair(rules, updates);
}

ints gen(const ints& lst, int idx)
{
    return ints(lst.end() - idx, lst.end());
}

int Compute(mmap& rules, std::vector<ints>& updates_list)
{
    std::set<ints> bad_updates {};
    std::set<ints> good_updates {};

    for (const auto& updates : updates_list) {
        for (std::size_t i { 0U }; i < updates.size(); ++i) {
            auto page = updates[updates.size() - 2 - i];
            auto elem = gen(updates, i + 1);

            for (const auto& e : elem) {
                auto it = rules.find(std::to_string(e));
                if (it != rules.end() && std::find(it->second.begin(), it->second.end(), page) != it->second.end())
                    bad_updates.insert(updates);
                else
                    good_updates.insert(updates);
            }
        }
    }

    for (const auto& updates : updates_list)
        good_updates.insert(updates);

    for (const auto& bad_update : bad_updates)
        good_updates.erase(bad_update);

    int sum = 0;
    for (const auto& good : good_updates)
        sum += good[good.size() / 2];

    return sum;
}

int main()
{
    auto [rules, updates] = Parse("input.txt");
    std::cout << Compute(rules, updates) << std::endl;

    return 0;
}
